src/turbine/turbine_tree.zig

```diff
@@ -0,0 +1,1100 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+
+const IpAddr = sig.net.IpAddr;
+const SocketAddr = sig.net.SocketAddr;
+const ShredId = sig.ledger.shred.ShredId;
+const RwMux = sig.sync.RwMux;
+const ThreadSafeContactInfo = sig.gossip.data.ThreadSafeContactInfo;
+const ContactInfo = sig.gossip.data.ContactInfo;
+const SignedGossipData = sig.gossip.data.SignedGossipData;
+const Pubkey = sig.core.Pubkey;
+const Epoch = sig.core.Epoch;
+const Slot = sig.core.Slot;
+const Duration = sig.time.Duration;
+const Instant = sig.time.Instant;
+const GossipTable = sig.gossip.GossipTable;
+const WeightedShuffle = sig.rand.WeightedShuffle(u64);
+const ChaChaRng = sig.rand.ChaChaRng(20);
+const AtomicUsize = std.atomic.Value(usize);
+const ThreadPool = sig.sync.ThreadPool;
+const KeyPair = std.crypto.sign.Ed25519.KeyPair;
+const SecretKey = std.crypto.sign.Ed25519.SecretKey;
+
+const intRangeLessThanRust = sig.rand.weighted_shuffle.intRangeLessThanRust;
+
+/// TurbineTreeCache
+/// Cache turbine trees and clear them once they are too old.
+/// The time to live ensures updates to gossip data are reflected
+/// in the turbine trees.
+pub const TurbineTreeCache = struct {
```

Comment:
This will leak memory across Epoch boundaries because it never evicts items until you try to get that specific item. But after an epoch transition you will stop getting the turbine tree from that epoch. So you'll keep around one turbine tree for every epoch forever.

You might want to use an Lru instead of a hashmap. You probably don't need more than two epochs at a time so you could potentially go simpler than an lru, like just an array of two items or something

Reply:
good catch, please see this simple fix and let me know if you are okay with it 
9d7741ec

Reply:
i like the simplicity of that fix. eventually though, a different design may be preferable

https://github.com/orgs/Syndica/projects/2/views/10?pane=issue&itemId=88567707